packages/gatsby/src/utils/page-data.ts

```diff
@@ -178,3 +180,50 @@ export function enqueueFlush(): void {
     flush()
   }
 }
+
+export async function handleStalePageData(): Promise<void> {
+  if (!(await fs.pathExists(`public/page-data`))) {
+    return
+  }
+
+  // public directory might have stale page-data files from previous builds
+  // we get the list of those and compare against expected page-data files
+  // and remove ones that shouldn't be there anymore
+
+  const activity = reporter.activityTimer(`Cleaning up stale page-data`)
+  activity.start()
+
+  const pageDataFilesFromPreviousBuilds = await new Promise<Set<string>>(
+    resolve => {
+      const results = new Set<string>()
+
+      const stream = fsWalkStream(`public/page-data`, {
+        deepFilter: entry => entry.path !== `public/page-data/sq/d`,
```

Comment:
Can you add a comment to explain what sq/d is and why it is special?

Reply:
So the `sq/d` is location for static query results, and because I construct list of expected files from pages it would remove those otherwise.

But now that I think about this - I already have additional check for `page-data.json` below and the `deepFilter` just tell `fsWalk` to not enter that directory at all. This can actually cause problems if user actually have pages under `/sq/d` "prefix", so I should remove that filter (let walker get into that directory) and rely just on filename check below. That was overzealous optim that will save 1 dir read at the cost of very weird potential edge cases.

Reply:
Handled in https://github.com/gatsbyjs/gatsby/pull/26937/commits/00fb3edde105906e4ceb6f39769d2c2a0b1d083a (removing it as special case)

Reply:
Less code is better. Nice!